File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/com/ririnto/sinon/harness/rules/HarnessCheckRule.java
package com.ririnto.sinon.harness.rules;

import com.ririnto.sinon.harness.Finding;
import com.ririnto.sinon.harness.core.RuleContext;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;

/**
 * Strategy interface implemented by each harness validation rule.
 */
public interface HarnessCheckRule {
    /**
     * Gets the manifest category key used by this rule.
     *
     * @return the manifest category key
     */
    String category();

    /**
     * Determines whether this rule applies to the given context.
     *
     * @param ctx shared rule context
     * @return true if this rule should be validated
     */
    boolean applies(RuleContext ctx);

    /**
     * Validates the given project directory against this rule.
     *
     * @param ctx shared rule context
     * @return a collection of findings; empty if validation passes
     * @throws MojoExecutionException if validation fails
     */
    Collection<Finding> validate(RuleContext ctx) throws MojoExecutionException;

    /**
     * Auto-formats the project against this rule, when supported.
     *
     * Rules without an automatic fix MUST keep this default and return an empty collection.
     *
     * @param ctx shared rule context
     * @return a collection of absolute paths modified by this rule
     * @throws MojoExecutionException if formatting fails
     */
    default Collection<Path> format(RuleContext ctx) throws MojoExecutionException {
        return List.of();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/com/ririnto/sinon/harness/rules/HarnessCheckRule.kt
package com.ririnto.sinon.harness.rules

import com.ririnto.sinon.harness.core.RuleContext
import com.ririnto.sinon.harness.ast.HarnessAstResults.Finding
import java.nio.file.Path

/**
 * Base class for harness validation rules.
 */
abstract class HarnessCheckRule {
    /**
     * Category key used in the harness manifest and findings.
     */
    abstract val category: String

    /**
     * Check whether this rule applies based on manifest configuration.
     *
     * @param ctx The rule context.
     * @return true if the rule applies, false otherwise.
     */
    open fun applies(ctx: RuleContext): Boolean = ctx.manifest.isEnabled(category)

    /**
     * Validate and return a collection of findings.
     *
     * @param ctx The rule context.
     * @return Collection of findings; empty collection if no issues found.
     */
    abstract fun validate(ctx: RuleContext): Collection<Finding>

    /**
     * Auto-format the project against this rule, when supported.
     *
     * Rules without an automatic fix MUST keep this default and return an empty collection.
     *
     * @param ctx The rule context.
     * @return Collection of absolute paths modified by this rule.
     */
    open fun format(ctx: RuleContext): Collection<Path> = emptyList()
}
